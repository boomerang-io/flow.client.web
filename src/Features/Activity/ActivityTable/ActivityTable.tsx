//@ts-nocheck
import React from "react";
import PropTypes from "prop-types";
import { settings } from "carbon-components";
import { DataTableSkeleton, DataTable, Pagination } from "@carbon/react";
import cx from "classnames";
import moment from "moment";
import queryString from "query-string";
import { getHumanizedDuration, isAccessibleKeyboardEvent } from "@boomerang-io/utils";
import EmptyState from "Components/EmptyState";
import { ExecutionStatusCopy, executionStatusIcon } from "Constants";
import styles from "./activityTable.module.scss";

const { prefix } = settings;

ActivityTable.propTypes = {
  history: PropTypes.object.isRequired,
  isLoading: PropTypes.bool,
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  tableData: PropTypes.object,
  updateHistorySearch: PropTypes.func.isRequired,
};

const PAGE_SIZES = [5, 10, 20, 25, 50, 100];

const headers = [
  {
    header: "Team",
    key: "teamName",
    sortable: true,
  },
  { header: "Scope", key: "scope" },
  {
    header: "Workflow",
    key: "workflowName",
    sortable: true,
  },
  {
    header: "Trigger",
    key: "trigger",
  },
  {
    header: "Initiated By",
    key: "initiatedByUserName",
  },
  {
    header: "Start Time",
    key: "creationDate",
  },
  {
    header: "Duration",
    key: "duration",
  },
  {
    header: "Status",
    key: "status",
  },
];

function renderCell(headerList, cellIndex, value) {
  const column = headerList[cellIndex];

  switch (column.key) {
    case "trigger":
      return (
        <p className={styles.tableTextarea} style={{ textTransform: "capitalize" }}>
          {value || "---"}
        </p>
      );
    case "creationDate":
      return <time className={styles.tableTextarea}>{moment(value).format("YYYY-MM-DD hh:mm A")}</time>;
    case "duration":
      return (
        <time className={styles.tableTextarea}>{value ? getHumanizedDuration(parseInt(value / 1000, 10)) : "---"}</time>
      );
    case "status":
      const Icon = executionStatusIcon[value ? value : "notstarted"];
      return (
        <div className={`${styles.status} ${styles[value]}`}>
          <Icon aria-label={value} className={styles.statusIcon} />
          <p className={styles.statusText}>{ExecutionStatusCopy[value ? value : "notstarted"]}</p>
        </div>
      );
    case "scope":
      return (
        <p className={styles.tableTextarea} style={{ textTransform: "capitalize" }}>
          {value || "---"}
        </p>
      );
    default:
      return <p className={styles.tableTextarea}>{value || "---"}</p>;
  }
}

function ActivityTable(props) {
  let headerList = headers;

  function handlePaginationChange({ page, pageSize }) {
    props.updateHistorySearch({
      ...queryString.parse(props.location.search),
      page: page - 1, // We have to decrement by one to offset the table pagination adjustment
      size: pageSize,
    });
  }

  function handleSort(e, { sortHeaderKey }) {
    const { property, direction } = props.tableData.pageable.sort[0];
    const sort = sortHeaderKey;
    let order = "ASC";

    if (sort === property && direction === "ASC") {
      order = "DESC";
    }

    props.updateHistorySearch({ ...queryString.parse(props.location.search), sort, order });
  }

  function executionViewRedirect(activityId) {
    const activity = props.tableData.records.find((activity) => activity.id === activityId);
    props.history.push({
      pathname: `/activity/${activity.workflowId}/execution/${activity.id}`,
      state: { fromUrl: `${props.match.url}${props.location.search}`, fromText: "Activity" },
    });
  }

  if (props.isLoading) {
    return (
      <div style={{ marginTop: "1rem" }}>
        <DataTableSkeleton
          className={cx(`${prefix}--skeleton`, `${prefix}--data-table`, styles.tableSkeleton)}
          rowCount={10}
          columnCount={headerList.length}
          headers={headerList.map((header) => header.header)}
        />
      </div>
    );
  }

  const {
    pageable: { number, size, sort, totalElements },
    records,
  } = props.tableData;
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;

  return (
    <section>
      <div className={styles.tableContainer}>
        {totalElements > 0 ? (
          <>
            <DataTable
              rows={records}
              headers={headerList}
              render={({ rows, headers, getHeaderProps }) => (
                <TableContainer>
                  <Table isSortable>
                    <TableHead>
                      <TableRow>
                        {headers.map((header) => (
                          <TableHeader
                            id={header.key}
                            {...getHeaderProps({
                              header,
                              className: `${styles.tableHeadHeader} ${styles[header.key]}`,
                              isSortable: header.sortable,
                              onClick: handleSort,
                            })}
                            isSortHeader={sort[0].property === header.key}
                            sortDirection={sort[0].direction}
                          >
                            {header.header}
                          </TableHeader>
                        ))}
                      </TableRow>
                    </TableHead>
                    <TableBody className={styles.tableBody}>
                      {rows.map((row) => (
                        <TableRow
                          key={row.id}
                          className={`${styles.tableRow} ${styles[row.cells[row.cells.length - 1].value]}`}
                          data-testid="configuration-property-table-row"
                          onClick={() => executionViewRedirect(row.id)}
                          onKeyDown={(e) => isAccessibleKeyboardEvent(e) && executionViewRedirect(row.id)}
                          tabIndex={0}
                        >
                          {row.cells.map((cell, cellIndex) => (
                            <TableCell key={cell.id}>
                              <div className={styles.tableCell}>{renderCell(headerList, cellIndex, cell.value)}</div>
                            </TableCell>
                          ))}
                        </TableRow>
                      ))}
                    </TableBody>
                  </Table>
                </TableContainer>
              )}
            />
            <Pagination
              onChange={handlePaginationChange}
              page={number + 1} // We need to offset by one bc service is zero indexed
              pageSize={size}
              pageSizes={PAGE_SIZES}
              totalItems={totalElements}
            />
          </>
        ) : (
          <>
            <DataTable
              rows={[]}
              headers={headerList}
              render={({ headers }) => (
                <TableContainer>
                  <Table>
                    <TableHead>
                      <TableRow className={styles.tableHeadRow}>
                        {headers.map((header) => (
                          <TableHeader
                            key={header.key}
                            id={header.key}
                            className={`${styles.tableHeadHeader} ${styles[header.key]}`}
                          >
                            {header.header}
                          </TableHeader>
                        ))}
                      </TableRow>
                    </TableHead>
                  </Table>
                </TableContainer>
              )}
            />
            <EmptyState message={"Execute some workflows or try changing the filters"} title="No activity found" />
          </>
        )}
      </div>
    </section>
  );
}

export default ActivityTable;
