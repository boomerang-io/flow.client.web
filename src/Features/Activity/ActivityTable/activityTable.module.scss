@import "Styles/mixins";

.tableSkeleton {
  th {
    color: var(--cds-interactive-01);
  }
  thead {
    background-color: #d5d9e0;
    th:nth-of-type(1) {
      width: 15% !important;
    }

    th:nth-of-type(2) {
      width: 25% !important;
    }

    th:nth-of-type(3) {
      width: 10% !important;
    }

    th:nth-of-type(4) {
      width: 15% !important;
    }

    th:nth-of-type(5) {
      width: 15% !important;
    }

    th:nth-of-type(6) {
      width: 10% !important;
    }

    th:nth-of-type(7) {
      width: 10% !important;
    }
  }
}

.tableContainer {
  margin-top: 1rem;
  margin-bottom: 2rem;
}

.header {
  display: flex;
  justify-content: space-between;
}

.tableHeadHeader {
  color: var(--cds-text-01);
  font-size: 0.875rem;
  font-weight: 600;
  height: 3rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;

  button {
    width: 100% !important;
  }

  &.teamName {
    width: 15%;
  }

  &.workflowName {
    width: 25%;
  }

  &.trigger {
    width: 10%;
  }

  &.initiatedByUserName {
    width: 15%;
  }

  &.creationDate {
    width: 15%;
  }

  &.duration {
    width: 10%;
  }

  &.status {
    width: 10%;
  }

  &.scope {
    width: 10%;
  }
}

.tableBody {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
}

.tableRow {
  cursor: pointer;

  &:focus {
    @include focus-outline(var(--cds-interactive-01), -0.125rem);
  }

  &.succeeded {
    .status {
      color: var(--cds-support-02);
      svg {
        fill: var(--cds-support-02);
      }
    }
  }

  &.failed, &.cancelled {
    td {
      background-color: #fcd0d3;
    }

    &:hover {
      background-color: #fcd0d3;
      td {
        background-color: #fcd0d3;
      }
    }

    .status {
      color: var(--flow-failure);
      svg {
        fill: var(--flow-failure);
      }
    }
  }

  &.invalid {
    td {
      background-color: #d5d9e0;
    }

    &:hover {
      background-color: #d5d9e0;
      td {
        background-color: #d5d9e0;
      }
    }
  }

  &.invalid,
  &.skipped {
    color: var(--cds-text-01);
    .status {
      color: var(--cds-text-01);
      svg {
        fill: var(--cds-text-01);
      }
    }
  }

  &.notstarted {
    .status {
      color: #50565b;
      svg {
        fill: #50565b;
      }
    }
  }

  &.running {
    .tableCell {
      animation: runningText 3.75s infinite;
      svg {
        animation: runningIcon 3.75s infinite;
      }
    }
  }
}

@keyframes runningText {
  0% {
    color: var(--cds-text-02);
  }
  20% {
    color: var(--cds-interactive-01);
  }
  40% {
    color: var(--cds-support-02);
  }
  60% {
    color: var(--cds-inverse-support-04);
  }
  80% {
    color: var(--cds-text-03);
  }
  100% {
    color: var(--cds-text-02);
  }
}

@keyframes runningIcon {
  0% {
    fill: var(--cds-text-02);
  }
  20% {
    fill: var(--cds-interactive-01);
  }
  40% {
    fill: var(--cds-support-02);
  }
  60% {
    fill: var(--cds-inverse-support-04);
  }
  80% {
    fill: var(--cds-text-03);
  }
  100% {
    fill: var(--cds-text-02);
  }
}

.tableCell {
  .status {
    align-items: center;
    display: flex;
    .statusIcon {
      height: 1rem;
      width: 1rem;
      margin-right: 0.5rem;
    }
  }
}

.tableTextarea {
  background: inherit;
  margin-left: 0.25rem;
  max-width: 100%;
  resize: none;
  text-align: left;
  white-space: normal;
  width: 90%;
  word-break: break-all;
}
