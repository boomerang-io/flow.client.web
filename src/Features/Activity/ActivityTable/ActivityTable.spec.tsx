import { vi } from "vitest";
import ActivityTable from "./index";

const props = {
  activities: [],
  isUpdating: false,
  sort: {},
  tableData: { records: [], pageable: { number: 0, size: 10, sort: "asc", totalElements: 10 } },
  match: {
    params: "testid",
  },
  location: {},
  history: {},
  updateHistorySearch: vi.fn(),
};

describe("ActivityTable --- Snapshot", () => {
  it("Capturing Snapshot of ActivityTable", () => {
    const { baseElement } = global.rtlRouterRender(<ActivityTable {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
