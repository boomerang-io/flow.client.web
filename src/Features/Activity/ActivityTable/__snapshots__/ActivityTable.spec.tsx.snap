// Vitest Snapshot v1

exports[`ActivityTable --- Snapshot > Capturing Snapshot of ActivityTable 1`] = `
<body
  id="app"
>
  <div>
    <section>
      <div
        class="_tableContainer_8eafaf"
      >
        <div
          class="cds--data-table-container"
        >
          <div
            class="cds--data-table-content"
          >
            <table
              class="cds--data-table cds--data-table--sort"
            >
              <thead>
                <tr>
                  <th
                    aria-sort="none"
                    class="_tableHeadHeader_8eafaf _teamName_8eafaf"
                    id="teamName"
                    scope="col"
                  >
                    <div
                      id="table-sort-1"
                      style="display: none;"
                    >
                      Click to sort rows by Team header in ascending order
                    </div>
                    <button
                      aria-describedby="table-sort-1"
                      class="_tableHeadHeader_8eafaf _teamName_8eafaf cds--table-sort"
                      type="button"
                    >
                      <span
                        class="cds--table-sort__flex"
                      >
                        <div
                          class="cds--table-header-label"
                        >
                          Team
                        </div>
                        <svg
                          aria-hidden="true"
                          class="cds--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="cds--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _scope_8eafaf"
                    id="scope"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Scope
                    </div>
                  </th>
                  <th
                    aria-sort="none"
                    class="_tableHeadHeader_8eafaf _workflowName_8eafaf"
                    id="workflowName"
                    scope="col"
                  >
                    <div
                      id="table-sort-3"
                      style="display: none;"
                    >
                      Click to sort rows by Workflow header in ascending order
                    </div>
                    <button
                      aria-describedby="table-sort-3"
                      class="_tableHeadHeader_8eafaf _workflowName_8eafaf cds--table-sort"
                      type="button"
                    >
                      <span
                        class="cds--table-sort__flex"
                      >
                        <div
                          class="cds--table-header-label"
                        >
                          Workflow
                        </div>
                        <svg
                          aria-hidden="true"
                          class="cds--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="cds--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _trigger_8eafaf"
                    id="trigger"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Trigger
                    </div>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _initiatedByUserName_8eafaf"
                    id="initiatedByUserName"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Initiated By
                    </div>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _creationDate_8eafaf"
                    id="creationDate"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Start Time
                    </div>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _duration_8eafaf"
                    id="duration"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Duration
                    </div>
                  </th>
                  <th
                    class="_tableHeadHeader_8eafaf _status_8eafaf"
                    id="status"
                    scope="col"
                  >
                    <div
                      class="cds--table-header-label"
                    >
                      Status
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                aria-live="polite"
                class="_tableBody_8eafaf"
              />
            </table>
          </div>
        </div>
        <div
          class="cds--pagination cds--pagination--md"
        >
          <div
            class="cds--pagination__left"
          >
            <label
              class="cds--pagination__text"
              for="cds-pagination-select-id-14"
              id="cds-pagination-select-id-14-count-label"
            >
              Items per page:
            </label>
            <div
              class="cds--form-item cds--select__item-count"
            >
              <div
                class="cds--select cds--select--inline"
              >
                <div
                  class="cds--select-input--inline__wrapper"
                >
                  <div
                    class="cds--select-input__wrapper"
                  >
                    <select
                      class="cds--select-input"
                      id="cds-pagination-select-id-14"
                    >
                      <option
                        class="cds--select-option"
                        value="5"
                      >
                        5
                      </option>
                      <option
                        class="cds--select-option"
                        value="10"
                      >
                        10
                      </option>
                      <option
                        class="cds--select-option"
                        value="20"
                      >
                        20
                      </option>
                      <option
                        class="cds--select-option"
                        value="25"
                      >
                        25
                      </option>
                      <option
                        class="cds--select-option"
                        value="50"
                      >
                        50
                      </option>
                      <option
                        class="cds--select-option"
                        value="100"
                      >
                        100
                      </option>
                    </select>
                    <svg
                      aria-hidden="true"
                      class="cds--select__arrow"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="cds--pagination__text cds--pagination__items-count"
            >
              1–10 of 10 items
            </span>
          </div>
          <div
            class="cds--pagination__right"
          >
            <div
              class="cds--form-item cds--select__page-number"
            >
              <div
                class="cds--select cds--select--inline"
              >
                <label
                  class="cds--label cds--visually-hidden"
                  for="cds-pagination-select-id-14-right"
                >
                  Page number, of 1 pages
                </label>
                <div
                  class="cds--select-input--inline__wrapper"
                >
                  <div
                    class="cds--select-input__wrapper"
                  >
                    <select
                      class="cds--select-input"
                      id="cds-pagination-select-id-14-right"
                    >
                      <option
                        class="cds--select-option"
                        value="1"
                      >
                        1
                      </option>
                    </select>
                    <svg
                      aria-hidden="true"
                      class="cds--select__arrow"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="cds--pagination__text"
            >
              of 1 page
            </span>
            <div
              class="cds--pagination__control-buttons"
            >
              <span
                class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip"
              >
                <button
                  aria-labelledby="tooltip-11"
                  class="cds--btn--icon-only cds--pagination__button cds--pagination__button--backward cds--pagination__button--no-index cds--btn cds--btn--md cds--btn--ghost cds--btn--disabled"
                  disabled=""
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20 24L10 16 20 8z"
                    />
                  </svg>
                </button>
                <span
                  aria-hidden="true"
                  class="cds--popover"
                  id="tooltip-11"
                  role="tooltip"
                >
                  <span
                    class="cds--popover-content cds--tooltip-content"
                  >
                    Previous page
                  </span>
                  <span
                    class="cds--popover-caret"
                  />
                </span>
              </span>
              <span
                class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top-right cds--tooltip cds--icon-tooltip"
              >
                <button
                  aria-labelledby="tooltip-13"
                  class="cds--btn--icon-only cds--pagination__button cds--pagination__button--forward cds--pagination__button--no-index cds--btn cds--btn--md cds--btn--ghost cds--btn--disabled"
                  disabled=""
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 8L22 16 12 24z"
                    />
                  </svg>
                </button>
                <span
                  aria-hidden="true"
                  class="cds--popover"
                  id="tooltip-13"
                  role="tooltip"
                >
                  <span
                    class="cds--popover-content cds--tooltip-content"
                  >
                    Next page
                  </span>
                  <span
                    class="cds--popover-caret"
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
